src/neo/Network/P2P/SyncManager.cs

```diff
@@ -0,0 +1,261 @@
+    {
+        public class Register { public VersionPayload Version; }
+        public class Task { public uint StartIndex; public uint EndIndex; public BitArray IndexArray; public DateTime Time; }
+        public class PersistedBlockIndex { public uint PersistedIndex; }
+        public class InvalidBlockIndex { public uint InvalidIndex; }
+        public class Update { public uint LastBlockIndex; }
+        private class Timer { }
+
+        private static readonly TimeSpan TimerInterval = TimeSpan.FromSeconds(30);
+        private static readonly TimeSpan SyncTimeout = TimeSpan.FromMinutes(1);
+
+        private readonly Dictionary<IActorRef, SyncSession> sessions = new Dictionary<IActorRef, SyncSession>();
+        private readonly NeoSystem system;
+        private readonly ICancelable timer = Context.System.Scheduler.ScheduleTellRepeatedlyCancelable(TimerInterval, TimerInterval, Context.Self, new Timer(), ActorRefs.NoSender);
+        private readonly List<Task> uncompletedTasks = new List<Task>();
+
+        private const uint BlocksPerTask = 50;
+        private const int MaxTasksCount = 10;
+
+        private readonly int maxTasksPerSession = 3;
+        private int totalTasksCount = 0;
+        private uint highestBlockIndex = 0;
+        private uint lastTaskIndex = 0;
+        private uint persistIndex = 0;
+
+        public SyncManager(NeoSystem system)
+        {
+            this.system = system;
+        }
+
+        protected override void OnReceive(object message)
+        {
+            switch (message)
+            {
+                case Register register:
+                    OnRegister(register.Version);
+                    break;
+                case Block block:
+                    OnReceiveBlock(block);
+                    break;
+                case PersistedBlockIndex blockIndex:
+                    OnReceiveBlockIndex(blockIndex);
+                    break;
+                case InvalidBlockIndex invalidBlockIndex:
+                    OnReceiveInvalidBlockIndex(invalidBlockIndex);
+                    break;
+                case Timer _:
+                    OnTimer();
+                    break;
+                case Update update:
+                    OnUpdate(update.LastBlockIndex);
+                    break;
+                case Terminated terminated:
+                    OnTerminated(terminated.ActorRef);
+                    break;
+            }
+        }
+
+        private void OnReceiveInvalidBlockIndex(InvalidBlockIndex invalidBlockIndex)
+        {
+            if (sessions.Count() == 0 || sessions.Values.Where(p => p.HasTask).Count() == 0) return;
+            foreach (SyncSession session in sessions.Values.Where(p => p.HasTask))
+            {
+                for (int i = 0; i < session.Tasks.Count(); i++)
+                {
+                    if (session.Tasks[i].StartIndex <= invalidBlockIndex.InvalidIndex && session.Tasks[i].EndIndex >= invalidBlockIndex.InvalidIndex)
```

Comment:
We need to check if there are more Unverified Candidates to be processed, maybe it was just a spam.

Reply:
We could deal with it in the next phase. Maintain the health of the remotenodes to avoid attacks.

Reply:
@ShawnYun, can or open an issue with that? Or should I open?
I prefer if you open and keep taking a look at this, you can count with me support.

Reply:
@vncoelho Thanks, I will open a issue about it.

Reply:
@vncoelho We can talk about this in #1316